pages/admin_pages.py

```diff
@@ -193,3 +199,46 @@ def reset_download_api_credentials():
         % (access_key, secret_key)
     flash(Markup(msg), 'warning')
     return redirect("/manage_credentials")
+
+
+@admin_pages.route('/new_api_key', methods=['POST'])
+@authenticate_researcher_login
+def new_api_key():
+    tableau_api_permission = 'api_permission' in request.values
+    readable_name = request.values.get("readable_name", "")
+    researcher = Researcher.objects.get(username=session[SESSION_NAME])
+    api_key = ApiKey.generate(researcher=researcher, has_tableau_api_permissions=tableau_api_permission, readable_name=readable_name)
+    msg = """<h3>New Data-Download API credentials have been generated for you!</h3>
+        <p>Your new <b>Access Key</b> is:
+          <div class="container-fluid">
+            <textarea rows="1" cols="85" readonly="readonly" onclick="this.focus();this.select()">%s</textarea></p>
+          </div>
+        <p>Your new <b>Secret Key</b> is:
+          <div class="container-fluid">
+            <textarea rows="1" cols="85" readonly="readonly" onclick="this.focus();this.select()">%s</textarea></p>
+          </div>
+        <p>Please record these somewhere; This secret key will not be shown again!</p>""" \
+          % (api_key.access_key_id, api_key.access_key_secret_plaintext)
+    api_key.save()
+    flash(Markup(msg), 'warning')
+    return redirect("/manage_credentials")
+
+
+@admin_pages.route('/disable_api_key', methods=['POST'])
+@authenticate_researcher_login
+def disable_api_key():
+    if "api_key_id" not in request.values:
+        flash(Markup("No API key specified"), 'warning')
+        return redirect("/manage_credentials")
+    api_key_id = request.values["api_key_id"]
+    api_key_query = ApiKey.objects.filter(access_key_id=api_key_id).filter(researcher=get_session_researcher())
+    if not api_key_query.exists():
+        flash(Markup("No matching API key found to disable"), 'warning')
+        return redirect("/manage_credentials")
+    api_key = api_key_query[0]
+    if not api_key.is_active:
+        flash("That API key has already been disabled: %s" % api_key_id, 'warning')
+        return redirect("/manage_credentials")
+    api_key.is_active = False
+    api_key.save()
+    return redirect("/manage_credentials")
```

Comment:
Can we formulate this as a `DELETE` request?

Reply:
When researching how to do this, I found out that HTML5 explicitly does not support methods other than GET and POST in forms. This [stack exchange thread] (https://softwareengineering.stackexchange.com/questions/114156/why-are-there-are-no-put-and-delete-methods-on-html-forms) goes into it, but it seems like a contentious topic with a some history (at one point it was implemented, then removed). There are ways to do it, but it seems to me that the design intention for HTML is for deletions to be done with POST requests and would be more inclined to leave it as is. What is your opinion?

Reply:
Hmmm, I think this point resonates most with me:
```
HTML is inextricably bound to HTTP. HTML is the human interface of HTTP. It's therefore automatically questionable why HTML does not support all relevant methods in the HTTP specification. Why can machines PUT and DELETE resources, but humans cannot?
```

So my take would be to keep the `DELETE` request on the backend and just switch to using a JS request to handle deletes (i.e. `Executing PUT and DELETE to modify resources on the origin server is straight-forward for modern Web browsers using the XmlHttpRequest object`), though the notable problem with this is if a browser has JS turned off.

However, the other complication is doing it this way means you also need to handle injecting the CSRF token into the JS request, which can range from fairly easy to not very trivial.

As much as I want to keep the semantics of using a `DELETE` request, I think the easier solution would be to keep the endpoint as something like `api-key/delete` and just use `POST`.